forem/forem
# Only permit valid class names as sponsorable type

## What type of PR is this? (check all applicable)


## Description

https://github.com/forem/forem/issues/14386 identified an error could
arise if input was provide to the sponsorable_type field in the admin
creation form, but it was not a valid constant (since we include the
related model in the index when loading @sponsorships, this permits
creating a sponsorship that can't easily be managed or deleted).

Add a validation to ensure when the sponsorable type is present, and it's
a sponsorable class. Currently that's `Tag` and it's ancestor `ActsAsTaggableOn::Tag`.

A "worse is better" solution would be to remove the polymorphic type from the admin form, and select tags directly (by selecting a name) rather than asking for an id as text. If we ever need to support sponsorship for non-tags, the UI would be updated. A less drastic change to the UI could be to make a dropdown from the `SPONSORABLE_TYPES` constant (while still requiring an id as well as a type).

I think any UI updates could be separated into an issue for contribution if that's something we want to do.

## Related Tickets & Documents

#14386 

https://github.com/forem/forem/pull/3438 - particularly the comments about why a polymorphic association was chosen, this reinforces the idea that "Tag" and "ActsAsTaggableOn::Tag" are the only currently valid constants (and the goal was to support future options rather than any specific need currently)

## QA Instructions, Screenshots, Recordings

fill in the form at http://localhost:3000/admin/advanced/sponsorships/new

for Sponsorable, choose 1 (does not matter)
for Sponsorable Type, choose any nonsense string (the unit test calls for "Unsponsorable", the reporter used "?c")

On form submission, validation should fail, record is not persisted, UI displays "Sponsorable type is not a valid class name" error message above form.

### UI accessibility concerns?

None

## Added/updated tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)
